pyop2/types/dataset.py

```diff
@@ -29,12 +29,19 @@ def __init__(self, iter_set, dim=1, name=None):
             return
         if isinstance(iter_set, Subset):
             raise NotImplementedError("Deriving a DataSet from a Subset is unsupported")
+        self.comm = mpi.internal_comm(iter_set.comm)
         self._set = iter_set
         self._dim = utils.as_tuple(dim, numbers.Integral)
         self._cdim = np.prod(self._dim).item()
         self._name = name or "dset_#x%x" % id(self)
         self._initialized = True
 
+    def __del__(self):
+        # Cannot use hasattr here, since we define `__getattr__`
+        # This causes infinite recursion when looked up!
+        if "comm" in self.__dict__:
```

Comment:
Maybe `try` and `except` would be cleaner?

Reply:
I think this will still hit the recursion issue when accessing `self.comm`